core/txpool/bundlepool/bundlepool.go

```diff
@@ -0,0 +1,439 @@
+	defer p.wg.Done()
+
+	for {
+		select {
+		case <-p.shutdownCh:
+			return
+		}
+	}
+}
+
+func (p *BundlePool) FilterBundle(bundle *types.Bundle) bool {
+	for _, tx := range bundle.Txs {
+		if !p.filter(tx) {
+			return false
+		}
+	}
+	return true
+}
+
+// AddBundle adds a mev bundle to the pool
+func (p *BundlePool) AddBundle(bundle *types.Bundle) error {
+	if p.simulator == nil {
+		return txpool.ErrSimulatorMissing
+	}
+
+	price, err := p.simulator.SimulateBundle(bundle)
+	if err != nil {
+		return err
+	}
+	minimalGasPrice := p.bundleGasPricer.MinimalBundleGasPrice()
+	if price.Cmp(minimalGasPrice) < 0 {
+		return txpool.ErrBundleGasPriceLow
+	}
+	bundle.Price = price
+
+	hash := bundle.Hash()
+	if _, ok := p.bundles[hash]; ok {
+		return txpool.ErrBundleAlreadyExist
+	}
+	for p.slots+numSlots(bundle) > p.config.GlobalSlots {
+		p.drop()
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.bundles[hash] = bundle
+	p.slots += numSlots(bundle)
+
+	bundleGauge.Update(int64(len(p.bundles)))
+	slotsGauge.Update(int64(p.slots))
+	return nil
+}
+
+func (p *BundlePool) GetBundle(hash common.Hash) *types.Bundle {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	return p.bundles[hash]
+}
+
+func (p *BundlePool) PruneBundle(hash common.Hash) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.slots -= numSlots(p.bundles[hash])
+	delete(p.bundles, hash)
+}
+
+func (p *BundlePool) PendingBundles(blockNumber *big.Int, blockTimestamp uint64) []*types.Bundle {
+	ret := make([]*types.Bundle, 0)
+
+	for hash, bundle := range p.bundles {
+		// Prune outdated bundles
+		if (bundle.MaxTimestamp != 0 && blockTimestamp > bundle.MaxTimestamp) ||
+			blockNumber.Cmp(new(big.Int).SetUint64(bundle.MaxBlockNumber)) > 0 {
+			p.PruneBundle(hash)
+			continue
+		}
+
+		// Roll over future bundles
+		if bundle.MinTimestamp != 0 && blockTimestamp < bundle.MinTimestamp {
+			continue
+		}
+
+		// return the ones that are in time
+		ret = append(ret, bundle)
+	}
+
+	bundleGauge.Update(int64(len(p.bundles)))
+	slotsGauge.Update(int64(p.slots))
+	return ret
+}
+
+// AllBundles returns all the bundles currently in the pool
+func (p *BundlePool) AllBundles() []*types.Bundle {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	bundles := make([]*types.Bundle, 0, len(p.bundles))
+	for _, bundle := range p.bundles {
+		bundles = append(bundles, bundle)
+	}
+	return bundles
+}
+
+func (p *BundlePool) Filter(tx *types.Transaction) bool {
+	return false
+}
+
+func (p *BundlePool) Close() error {
+	close(p.shutdownCh)
+	p.wg.Wait()
+
+	log.Info("Bundle pool stopped")
+	return nil
+}
+
+func (p *BundlePool) Reset(oldHead, newHead *types.Header) {
+	// TODO implement me
+	panic("implement me")
+}
+
+// SetGasTip updates the minimum price required by the subpool for a new
+// transaction, and drops all transactions below this threshold.
+func (p *BundlePool) SetGasTip(tip *big.Int) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	old := p.gasTip.Load()
+	p.gasTip.Store(new(big.Int).Set(tip))
+
+	if tip.Cmp(old) > 0 {
+		// TODO
```

Comment:
On BSC, the minimal gastip is fixed. Please consider whether this is needed.

Reply:
Removed